### ISSUE TITLE
Unravel vector expressions into for loops with indices where necessary

### ISSUE THREAD

**Issue by EmilyBourne** [pyccel/pyccel#695]
**Describe the bug**
In some languages vector expressions need to be expressed in loops. In C this is the case for all vector expressions, while in Fortran it is the case for expressions with mixed orders

**Test Code**
Provide code which does not currently work but which should do when this issue is fixed:
```python
import numpy as np
a = np.ones((3,4), order='C')
b = np.ones((3,4), order='F')
c = a + b
```

In fortran:
```fortran
program prog_tmp

use, intrinsic :: ISO_C_BINDING

implicit none

real(C_DOUBLE), allocatable :: a(:,:)
real(C_DOUBLE), allocatable :: b(:,:)
real(C_DOUBLE), allocatable :: c(:,:)
integer(C_INT) :: tmp_1
integer(C_INT) :: tmp_2

allocate(a(0:4_C_INT64_T - 1_C_INT64_T, 0:3_C_INT64_T - 1_C_INT64_T))
a = 1.0_C_DOUBLE
allocate(b(0:3_C_INT64_T - 1_C_INT64_T, 0:4_C_INT64_T - 1_C_INT64_T))
b = 1.0_C_DOUBLE
allocate(c(0:4_C_INT64_T - 1_C_INT64_T, 0:3_C_INT64_T - 1_C_INT64_T))

do tmp_1 = 0:3
  do tmp_2 = 0:3
     c(tmp_1, tmp_2) = a(tmp_1, tmp_2) + b(tmp_2, tmp_1)
  end do
end do

end program prog_tmp
```

Ideally the solution should satisfy the following criteria:
- [ ] Only add indices if necessary
    - Standard vector expressions supported by the Fortran language do not need indices
    - Mixed rank expressions in Fortran do not need indices for all dimensions. E.g. `A+b` where `A` is a matrix, and `b` is a vector can be handled with only 1 index
- [ ] Consecutive expressions requiring the same indices without dependencies should be grouped in the same for loop

**Comment by EmilyBourne** [comment]
This was treated in PR #623 but suggested improvements mean this issue is now blocked by issue #694

**Comment by EmilyBourne** [comment]
Fixed in PR #623
